File: web/src/reducers/apps.ts
/*
MIT License

Copyright (c) 2020 The KubeLens Authors

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
import { Reducer } from 'redux';
import { AppsActions, AppsActionTypes } from '../actions/apps';
import { AppOverview, App } from '../types';

export interface IAppsState {
  readonly appOverview?: AppOverview,
  readonly apps?: App[],
  readonly selectedAppName?: string,
  readonly filteredApps?: App[]
}

export const INITIAL_STATE: IAppsState = {
  appOverview: undefined,
  apps: undefined,
  selectedAppName: undefined,
  filteredApps: undefined
};

export const reducer: Reducer<IAppsState, AppsActions> = (
  state = INITIAL_STATE,
  action
) => {
  switch (action.type) {
    case AppsActionTypes.GET_APP_OVERVIEW: {
      return {
        ...state,
        appOverview: action.appOverview
      };
    }

    case AppsActionTypes.GET_APPS: {
      return {
        ...state,
        apps: action.apps,
        filteredApps: action.apps
      }
    }

    case AppsActionTypes.SET_SELECTED_APP_NAME: {
      return {
        ...state,
        selectedAppName: action.selectedAppName
      }
    }

    case AppsActionTypes.FILTER_APPS: {
      return {
        ...state,
        filteredApps: action.filteredApps
      }
    }

    default:
      return state;
  }
};

File: api/svc/apps.go
/*
MIT License

Copyright (c) 2020 The KubeLens Authors

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/

package svc

import (
	"encoding/json"
	"net/http"
	"strings"

	"github.com/kubelens/kubelens/api/auth/rbac"
	"github.com/kubelens/kubelens/api/errs"
	k8sv1 "github.com/kubelens/kubelens/api/k8sv1"

	"github.com/creack/httpreq"
	klog "github.com/kubelens/kubelens/api/log"
)

// Apps retrieves a list of applications.
func (h request) Apps(w http.ResponseWriter, r *http.Request) {
	l := klog.MustFromContext(r.Context())
	ra := rbac.MustFromContext(r.Context())

	apps, apiErr := h.k8Client.Apps(k8sv1.AppOptions{
		UserRole: ra,
		Logger:   l,
	})

	if apiErr != nil {
		l.Error(apiErr)
		http.Error(w, apiErr.Message, apiErr.Code)
		return
	}

	res, err := json.Marshal(appsResponse(apps))

	if err != nil {
		e := errs.SerializationError(err.Error())
		http.Error(w, e.Message, e.Code)
		return
	}

	w.WriteHeader(http.StatusOK)
	w.Write(res)
}

// AppOverview retrieves a list of applications overviews.
func (h request) AppOverview(w http.ResponseWriter, r *http.Request) {
	l := klog.MustFromContext(r.Context())
	ra := rbac.MustFromContext(r.Context())

	var appname string

	// "/v1/apps/{name}" = []string{"", "apps", "name"}
	if params := strings.Split(r.URL.Path, "/"); len(params) > 2 {
		appname = params[2]
	}

	// get query params
	var data Req
	if err := httpreq.NewParsingMapPre(1).
		ToString("namespace", &data.Namespace).
		ToString("labelSelector", &data.LabelSelector).
		ToBool("detailed", &data.Detailed).
		Parse(r.URL.Query()); err != nil {
		l.Error(err.Error())
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	overviews, apiErr := h.k8Client.AppOverview(k8sv1.AppOverviewOptions{
		UserRole:      ra,
		Logger:        l,
		AppName:       appname,
		LabelSelector: data.LabelSelectorMap(),
		Namespace:     data.Namespace,
		Detailed:      data.Detailed,
	})

	if apiErr != nil {
		l.Error(apiErr)
		http.Error(w, apiErr.Message, apiErr.Code)
		return
	}

	res, err := json.Marshal(overviews)

	if err != nil {
		e := errs.SerializationError(err.Error())
		http.Error(w, e.Message, e.Code)
		return
	}

	w.WriteHeader(http.StatusOK)
	w.Write(res)
}
